File: 162.Find-Peak-Element.java
// https://leetcode.com/problems/find-peak-element/
//
// algorithms
// Medium (41.15%)
// Total Accepted:    233,327
// Total Submissions: 567,038
// beats 100.0% of java submissions

class Solution {
    public int findPeakElement(int[] nums) {
        int len = nums.length;
        
        int l = 0, r = len - 1, m = -1;
        
        while (l < r) {
            m = l + (r - l) / 2;
            if (nums[m] > nums[m + 1]) {
                r = m;
            } else {
                l = m + 1;
            }
        }
        
        return l;
    }
}

File: 162.Find-Peak-Element.py
# https://leetcode.com/problems/intersection-of-two-linked-lists/
#
# algorithms
# Medium (41.75%)
# Total Accepted:    262,503
# Total Submissions: 628,724
# beats 100.0% of python submissions


class Solution(object):
    def findPeakElement(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        length = len(nums)
        l, r, m = 0, length - 1, -1

        while l <= r:
            m = l + (r - l) / 2
            pre = float('-inf') if m == 0 else nums[m - 1]
            next = float('-inf') if m == length - 1 else nums[m + 1]

            if nums[m] > pre and nums[m] > next:
                return m

            elif next > nums[m]:
                l = m + 1

            else:
                r = m - 1

        return -1
